### DISCUSSION TITLE
Release 3.4.0 - Line graphs & bugfix

### DISCUSSION THREAD

**Discussion by mathoudebine** [mathoudebine/turing-smart-screen-python#472] [Announcements]
## New feature: line graphs!
Thanks to @hicwic's PR #371 the System Monitor program now supports line graphs!
![image](https://github.com/mathoudebine/turing-smart-screen-python/assets/38615348/71bea6cb-6412-4b46-912e-6e571fa37eb9)

The line graph feature has been added to many metrics like: CPU/GPU/Disk/Memory usage, temperature, CPU Fan speed, FPS...
Check https://github.com/mathoudebine/turing-smart-screen-python/blob/main/res/themes/theme_example.yaml and look for `LINE_GRAPH` key:
```yaml
      LINE_GRAPH:
        SHOW: False
        X: 300
        Y: 220
        WIDTH: 133
        HEIGHT: 70
        MIN_VALUE: 0
        MAX_VALUE: 100
        HISTORY_SIZE: 10
        AUTOSCALE: False
        LINE_COLOR: 61, 184, 225
        AXIS: True
        AXIS_COLOR: 255, 135, 0
        # BACKGROUND_COLOR: 0, 0, 0
        BACKGROUND_IMAGE: background.png
```
The history size (number of previous values displayed on the graph) can be adjusted for its data.
With `AUTOSCALE: True` the line graph Y axis will constantly adjust to the min/max displayed values, otherwise set `MIN_VALUE`/ `MAX_VALUE` for a static Y axis
Graph axis can be show or hidden with `AXIS: True` / `False` and set to a specific color with `AXIS_COLOR` (default to the `LINE_COLOR` if not set)

## New themes:
This release comes with 2 new themes that use the line graph functionality:

### Cyberdeck theme (3.5 inch)
@hicwic released the theme he made for his Cyberdeck project
<img src="https://github.com/mathoudebine/turing-smart-screen-python/blob/main/res/themes/Cyberdeck/preview.png?raw=true"/>

Check this discussion #374 for more details and final prototype!
![282247624-9716c07c-f939-45af-8369-f067f6df4e11](https://github.com/mathoudebine/turing-smart-screen-python/assets/38615348/d16fcd9f-b73b-4556-b5b3-aaff4d059dc3)

### Task Manager theme (5 inch)
This theme takes full advantage of the line graph feature, and is inspired by the (in)famous Windows Task Manager
![theme_taskmanager5inch](https://github.com/mathoudebine/turing-smart-screen-python/assets/38615348/7cce302e-48ae-44e8-ba25-fe123afd58ce)

## Other fixes
This release also contain the following fixes:
- System Monitor: Workaround for models failing to display their previous content after being turned off/on #427 (fixes #419)
- Theme Editor: only refresh theme if theme file still exists (fix for vim text editor) #428 (fixes #401)
- System Monitor: Custom stats not displayed if numeric value is zero #444 (fixes #442)
- System Monitor: Better error message when missing ImageTk dependency #443
- System Monitor: Do not use psutil virtual_memory used/free: not reliable. Use total/available instead #462 (fixes #450)
- System Monitor: User proper conversion for Memory values (power of 1024 instead of power of 1000) #463 (fixes #449 partially)
- System Monitor: Fix LibreHardwareMonitor computation for swap/pagefile usage in case no swap/pagefile defined #391 (fixes #390)

## What's Changed
* Fix LibreHardwareMonitor computation for swap/pagefile usage in case no swap/pagefile defined by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/391
* Add listing of SubHardware from LibreHardwareMonitor by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/395
* :arrow_up: Update psutil requirement from ~=5.9.6 to ~=5.9.7 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/422
* :arrow_up: Bump actions/setup-python from 4 to 5 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/407
* :arrow_up: Bump actions/upload-artifact from 3 to 4 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/414
* :arrow_up: Bump github/codeql-action from 2 to 3 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/412
* :arrow_up: Update babel requirement from ~=2.13.1 to ~=2.14.0 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/411
* Workaround for models failing to display their previous content after being turned off/on by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/427
* Theme Editor: only refresh theme if theme file still exists (fix for vim text editor) by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/428
* Fix: Custom stats not displayed if numeric value is zero by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/444
* Update error message when missing ImageTk dependency by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/443
* :arrow_up: Update pillow requirement from ~=10.1.0 to ~=10.2.0 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/434
* :arrow_up: Update psutil requirement from ~=5.9.7 to ~=5.9.8 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/457
* :arrow_up: Bump actions/dependency-review-action from 3 to 4 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/454
* Feature - plot graph by @hicwic in https://github.com/mathoudebine/turing-smart-screen-python/pull/371
* Do not use psutil virtual_memory used/free: not reliable. Use total/available instead by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/462
* User proper conversion for Memory values (power of 1024 instead of power of 1000) by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/463
* :arrow_up: Update sv-ttk requirement from ~=2.5.5 to ~=2.6.0 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/464
* :arrow_up: Update ruamel-yaml requirement from ~=0.18.5 to ~=0.18.6 by @dependabot in https://github.com/mathoudebine/turing-smart-screen-python/pull/466
* @hicwic PR #371 by @mathoudebine in https://github.com/mathoudebine/turing-smart-screen-python/pull/461

## New Contributors
* @hicwic made their first contribution in https://github.com/mathoudebine/turing-smart-screen-python/pull/371

**Full Changelog**: https://github.com/mathoudebine/turing-smart-screen-python/compare/3.3.4...3.4.0

<hr /><em>This discussion was created from the release <a href='https://github.com/mathoudebine/turing-smart-screen-python/releases/tag/3.4.0'>Release 3.4.0 - Line graphs & bugfix</a>.</em>

**Comment by Rembo777** [comment]
Could you tell me whether this graph is able to display the actual frame time, allowing the current delays in milliseconds (‘ms’) to be monitored, similar to how MSI Afterburner + RTSS Rivatuner Statistics Server do it in the overlay?
